From c4f1e5134b8241ea37e46615fe3f48b5e8bd9b67 Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Sat, 28 May 2011 06:54:27 -0600
Subject: [PATCH 096/158] - cleanups with readdir: return "exact position
 found" for "." in cob iterator get method and handle this in mdstore readdir;
 - disregard nlink for dirs in mdstore unlink method (tmp hack for initial
 scan).

---
 cob/cob.c         | 17 +++++++++--------
 mdstore/mdstore.c | 14 +++++++++++---
 nrs/nrs.h         |  1 -
 3 files changed, 20 insertions(+), 12 deletions(-)

diff --git a/cob/cob.c b/cob/cob.c
index 0ec2171..fd23e09 100644
--- a/cob/cob.c
+++ b/cob/cob.c
@@ -645,20 +645,21 @@ int c2_cob_iterator_get(struct c2_cob_iterator *it)
 
         rc = c2_db_cursor_get(&it->ci_cursor, &it->ci_pair);
 
-        /* 
-         * Exact position found. 
+        /*
+         * Exact position found.
          */
-        if (rc == 0)
+        if (rc == 0 || (it->ci_key->cnk_name.b_len == 1 &&
+                        *(char *)it->ci_key->cnk_name.b_data == '.'))
                 return 1;
 
-        /* 
+        /*
          * Nothing found, cursor is on another object key. 
          */
         if (!c2_fid_eq(&it->ci_key->cnk_pfid, it->ci_cob->co_fid))
-                return -ENOENT;
-        
-        /* 
-         * Not exact position found. 
+                rc = -ENOENT;
+
+        /*
+         * Not exact position found.
          */
         return 0;
 }
diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index 5b21e90..b780a2a 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -253,9 +253,16 @@ int c2_md_store_unlink(struct c2_md_store       *md,
                         }
                 }
                 c2_free(nskey);
+                cob->co_nsrec.cnr_nlink--;
+        } else {
+                /*
+                 * We ignore nlink for dirs and go for killing it.
+                 * This is because we don't update parent nlink in
+                 * case of killing dirs.
+                 */
+                cob->co_nsrec.cnr_nlink = 0;
         }
 
-        cob->co_nsrec.cnr_nlink--;
         if (cob->co_nsrec.cnr_nlink == 0) {
                 /* 
                  * Last nlink, let's delete everything left. 
@@ -599,7 +606,7 @@ int c2_md_store_readdir(struct c2_md_store      *md,
         rc = c2_cob_iterator_get(&it);
         if (rc == 0) {
                 /* 
-                 * No exact position found and we are on least key, 
+                 * Not exact position found and we are on least key
                  * let's do one step forward.
                  */
                 rc = c2_cob_iterator_next(&it);
@@ -608,7 +615,7 @@ int c2_md_store_readdir(struct c2_md_store      *md,
     
         ent = rdpg->r_buf.b_addr;
         nob = rdpg->r_buf.b_nob;
-        while (rc == 0 || first || second) {
+        while (rc == 0) {
                 int next_step = 0;
                 
                 if (first) {
@@ -620,6 +627,7 @@ int c2_md_store_readdir(struct c2_md_store      *md,
                         name = "..";
                         len = 2;
                         second = 0;
+                        next_step = 1;
                 } else {
                         name = c2_bitstring_buf_get(&it.ci_key->cnk_name);
                         len = c2_bitstring_len_get(&it.ci_key->cnk_name);
diff --git a/nrs/nrs.h b/nrs/nrs.h
index 809a54e..390bc63 100644
--- a/nrs/nrs.h
+++ b/nrs/nrs.h
@@ -15,7 +15,6 @@ struct c2_cond;
 struct c2_mutex;
 struct c2_thread;
 struct c2_queue;
-
 struct c2_nrs_policy;
 
 struct c2_nrs {
-- 
1.8.3.2

